src/truss.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+
+  Copyright 2017 The Johns Hopkins University Applied Physics Laboratory LLC. All Rights Reserved.
+
+  Truss algorithm for cohesive subgroups.
+
+  Author: Alex Perrone
+  Date: 2017-08-03
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
+  02110-1301 USA
+
+*/
+
+#include "truss.h"
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ *
+ * \example examples/simple/igraph_truss.cpp
+ */
+int igraph_truss(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  // List the triangles as vertex triplets.
+  igraph_vector_int_init(&triangles, 0);  // will be resized
+  igraph_list_triangles(graph, &triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &triangles);
```

Comment:
`IGRAPH_FINALLY` must be called right after `igraph_vector_int_init` to ensure that igraph's memory management routine is already aware about the allocated memory for `triangles` even before `igraph_list_triangles()` is called (otherwise this memory would not be freed when `igraph_list_triangles()` returns with an error).